README.md
# zbx-smartctl

## Description

This is the template and discovery scripts for monitoring disks SMART attributes using smartmontools in Zabbix.  
Zabbix server 3.4+ is recommended with dependendent items support but there are also older templates for 3.2, 3.0, 2.4, 2.2 provided as is. Discovery scripts should work with them too.

### Main features

- Supports SATA, SAS and NVMe devices
- Disks discovery:
  - Two discovery scripts - for Linux/BSD/MacOS and Windows
  - Simple discovery in MacOS by scanning `/dev/disk/*` (macos)
  - Discover with smartctl --scan-open (nix, windows)
  - Discover NVMe devices with smartctl --scan-open -dnvme (nix, windows)
  - Discover Hardware RAID with `sg_scan` (nix only)
  - Discover NVMe devices with `nvme-cli` (nix only)
  - Handling usbjmicron (nix only)
  - Handling Areca SATA RAID (nix only)
  - Try to enable SMART if it is disabled(nix, macos, windows)
  - (new) static discovery (nix only)
  - HDD(0), SSD/NVMe(1), `other`(2) classification in {#DISKTYPE} macro (nix, macos, windows)
  - LLD macros in output: {#DISKNAME}, {#DISKCMD}, {#DISKTYPE}, {#DISKMODEL}, {#DISKSN}. {#SMART_ENABLED} (nix, macos, windows)
- Templates:
  - For all templates:
    - Zabbix agent required with UserParameter
    - LLD discovery of disks
    - 'Problems first' approach. Collect items that can help to detect disk failures
    - Skip disks if SMART_ENABLED != 1
    - SATA devices support
  - 3.4+ template:
    - Two discovery rules: for HDD and SSD/NVMe to reduce the number of unsupported
    - Server side regex parsing, so, very simple UserParameters in agent configs
    - No excessive calls to disks. Collect all items in the single smartctl run
    - SAS devices support
    - NVMe devices support
    - Static discovery support with {$SMARTCTL_STATIC_DISKS}

#### About static discovery

Static discovery is useful for disks that cannot be easily discovered, such as disks hidden behind some hardware RAIDs or when agent is installed on Windows or Mac where automatic discovery is not so powerful.

`{$SMARTCTL_STATIC_DISKS}` - If some disks cannot be discovered automatically no matter how hard you try, you can add additional disks with -d option in this macro on the host level. Such disks will be discovered in addition to any disks that will be discovered with smartctl --scan-open, sg_scan and so on.

Replace all spaces with `_` inside each disk command. Separate multiple disks with space ' '.
For example, to discover 2 drives behind hardware RAID, set this macro on the host level:

`{$SMARTCTL_STATIC_DISKS} = /dev/sda_-d_sat+megaraid,00 /dev/sda_-d_sat+megaraid,01`

## Installation

### Linux/BSD/Mac OSX

- Make sure that smartmontools package is installed
- (optional) Install `sg3-utils` if you need to monitor hardware RAIDs. See [#29](https://github.com/v-zhuravlev/zbx-smartctl/pull/29)
- (optional) Install `nvme-cli` if you need to monitor NVMe devices.  
- Copy the following contents of `sudoers_zabbix_smartctl` file to `/etc/sudoers.d/sudoers_zabbix_smartctl`:

```text
Cmnd_Alias SMARTCTL = /usr/sbin/smartctl
Cmnd_Alias SMARTCTL_DISCOVERY = /etc/zabbix/scripts/smartctl-disks-discovery.pl
zabbix ALL= (ALL) NOPASSWD: SMARTCTL, SMARTCTL_DISCOVERY
Defaults!SMARTCTL !logfile, !syslog, !pam_session
Defaults!SMARTCTL_DISCOVERY !logfile, !syslog, !pam_session
```

- Copy `zabbix_smartctl.conf` to `/etc/zabbix/zabbix_agentd.d`
- Copy script `smartctl-disks-discovery.pl` to `/etc/zabbix/scripts`
  - Then run
  
```text
chown zabbix:zabbix /etc/zabbix/scripts/smartctl-disks-discovery.pl
chmod u+x /etc/zabbix/scripts/smartctl-disks-discovery.pl
```

- Test the discovery script by running it as sudo. You should receive JSON object in the script output.
- Restart zabbix-agent


#### Building deb package

You can create .deb package `zabbix-agent-extra-smartctl` for Debian/Ubuntu distributions:

```shell
dpkg-buildpackage -tc -Zgzip
```

#### Ansible playbook

There is an ansible playbook available in this repo, feel free to try it.

### Windows

- Install [smartmontools](https://www.smartmontools.org/wiki/Download#InstalltheWindowspackage), prefer default installation path
- Install Zabbix agent using [official MSI package](https://www.zabbix.com/download_agents), prefer default installation path
- Copy script `smartctl-disks-discovery.ps1` to `Zabbix Agent` folder
- Copy file `zabbix_smartctl.win.conf` to `Zabbix Agent\zabbix_agentd.conf.d` folder
  - Check that path to smartmontools bin folder and to discovery script `smartctl-disks-discovery.ps1` are correct
- Restart Zabbix Agent service
- Test items
  - Test discovery and retrieval of disks data:

```text
PS C:\Program Files\Zabbix Agent> .\zabbix_agentd.exe -c .\zabbix_agentd.conf -t uHDD.discovery
uHDD.discovery                                [t|{
 "data":[
         {
                "{#DISKSN}":"ZZZZZZZZZZZZ",
                "{#DISKMODEL}":"THNSN5512GPUK TOSHIBA",
                "{#DISKNAME}":"/dev/sda",
                "{#DISKCMD}":"/dev/sda -dnvme",
                "{#SMART_ENABLED}":"0",
                "{#DISKTYPE}":"1"
         }
 ]
}]
PS C:\Program Files\Zabbix Agent> .\zabbix_agentd.exe -c .\zabbix_agentd.conf -t uHDD.get["/dev/sda -d nvme"]
uHDD.get[/dev/nvme1]                          [t|smartctl 6.6 2017-11-05 r4594 [x86_64-w64-mingw32-w10-b17134] (sf-6.6-1)..DISK OUTPUT HERE.....]
```


## Examples

Items in 3.4:
![image](https://user-images.githubusercontent.com/14870891/30012649-3b1cc758-914a-11e7-82d5-9c80abb1328f.png)
Catch SSD problems in Zabbix 2.4:
![image](https://user-images.githubusercontent.com/14870891/45471572-30d43580-b73a-11e8-8aa9-6d3260162ef3.png)

## Contributing

Please keep in mind key concepts when submitting a PR:

- The template should work with Windows, Linux, MacOS.
- Discovery scripts should not have any dependencies (apart from smartctl)
- Discovery scripts should deduplicate disks (using serial number as unique id)
- Discovery scripts should output the following set of macros:
  - {#DISKSN} - Disk serial number
  - {#DISKMODEL} - Disk model
  - {#DISKNAME} - Disk name you would like to use in item name
  - {#DISKCMD} - System disk name with -d param to be used in running smartctl
  - {#SMART_ENABLED} - 1 or 0
  - {#DISKTYPE} - 0 - HDD, 1 - SSD/NVMe, 2 - Other(ODD etc)
  
  To make sure that the sources of these macro is available everywhere, it is best to use output of `smartctl -i` or `smartctl --scan-open`. Other macros may be added, but try to edit both windows and nix scripts at the same time.

Please also keep in mind things that require improvement (welcome!)

- Absolute paths used(especially in Windows(UserParameters,inside powershell script))
- Discovery script should probably fail if not run under Admin/root(since its impossible to collect proper data)
- usbjmicron is not implemented in Windows, only in Linux discovery script
- nvme discovery is not implemented in Windows, only in Linux discovery script
- There are no proper tests. It's nice to run discovery scripts automatically using `/tests/examples` directory contents as mocks. So it's easier to accept PRs. BTW you can also PR your outputs to examples folder
- MacOS disks discovery is very limited. Feel free to improve it.

## Troubleshooting

1. SELinux. Turn it off or add a selinux policy:

```text
yum install policycoreutils-python
semanage permissive -a zabbix_agent_t
semodule -DB
# from zabbox_server issue:
zabbix_get -s host -k uHDD.discovery
zabbix_get -s host -k uHDD.get[/dev/sda]

cat /var/log/audit/audit.log | grep zabbix_agent_t | grep denied | audit2allow -M zabbix_smartctl
semodule -i zabbix_smartctl.pp
semanage permissive -d zabbix_agent_t
semodule -B
```

## License

GPL v3 or newer.

## More info

http://habrahabr.ru/company/zabbix/blog/196218/  
http://www.lanana.org/docs/device-list/devices-2.6+.txt  
https://www.smartmontools.org/wiki/Supported_RAID-Controllers  
https://www.percona.com/blog/2017/02/09/using-nvme-command-line-tools-to-check-nvme-flash-health/


debian/changelog
zabbix-extra-smartctl (0.1ubuntu4) UNRELEASED; urgency=low

  [ Alexey Vazhnov ]
  * Initial release.

  [ Alexander Lelyakin ]
  * Packaging corrected
  * add executable bit to script
  * postinst scripts added

  [ Vitaly Zhuravlev ]
  * improved sudoers file

 -- Vitaly Zhuravlev <zhuravlev.vitaly@gmail.com>  Fri, 19 Apr 2019 21:00:00 +0300


tests/test_attributes.py
import os
import re
import unittest
import xml.etree.ElementTree as ET


class TestCaseSmartAttributes(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        tree = ET.parse('Template_3.4_HDD_SMARTMONTOOLS_2_WITH_LLD.xml')

        # get files from dir
        cls.samples = []
        for (_, _, samples) in os.walk('tests/examples'):
            for sample in samples:
                with open('tests/examples/' + sample, 'r') as fd:
                    cls.samples.append({
                        "text": fd.read(),
                        "name": sample,
                        "disk_type": sample.split('_')[1],
                        "disk_interface": sample.split('_')[0]
                    })
        cls.root = tree.getroot()
        cls.text = """
smartctl 6.6 2017-11-05 r4594 [x86_64-w64-mingw32-w10-b17134] (sf-6.6-1)
Copyright (C) 2002-17, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Model Number:                       THNSN5512GPUK TOSHIBA
Serial Number:                      87PB60PKKSPX
Firmware Version:                   5KHA4102
PCI Vendor/Subsystem ID:            0x1179
IEEE OUI Identifier:                0x00080d
Controller ID:                      0
Number of Namespaces:               1
Namespace 1 Size/Capacity:          512 110 190 592 [512 GB]
Namespace 1 Formatted LBA Size:     512
Namespace 1 IEEE EUI-64:            00080d 03001d669b
Local Time is:                      Wed Jun 27 05:29:07 2018 RTZST
            """

    def get_regex_from_template(self, name):

        preprocessing_params = self.root.findall(
            ".//item_prototype[name='" + name + "']/preprocessing/step/params")[0].text
        (regex, group) = preprocessing_params.split("\n")
        group = int(group[1:])
        return (regex, group)

    def test_model(self):
        (regex, group) = self.get_regex_from_template(
            '{#DISKNAME}: Device model')
        m = re.search(regex, self.text)
        self.assertEqual(m.group(group), "THNSN5512GPUK TOSHIBA")

    def test_sata_attributes(self):

        attributes = [
            {'name': '{#DISKNAME}: ID 5 Reallocated Sectors Count', 'flags': []},
            {'name': '{#DISKNAME}: ID 9 Power On Hours', 'flags': []},
            {'name': '{#DISKNAME}: ID 10 Spin Retry Count', 'flags': ['hdd_only']},
            {'name': '{#DISKNAME}: ID 177/ID 202/ID 233 (SSD Wearout)', 'flags': ['ssd_only']},
            {'name': '{#DISKNAME}: ID 190 Airflow Temperature', 'flags': []},
            {'name': '{#DISKNAME}: ID 194 Temperature', 'flags': []},
            {'name': '{#DISKNAME}: ID 197 Current Pending Sector Count', 'flags': ['hdd_only']},
            {'name': '{#DISKNAME}: ID 198 (Offline) Uncorrectable Sector Count', 'flags': ['hdd_only']},
            {'name': '{#DISKNAME}: ID 199 UltraDMA CRC Error Count', 'flags': []}
        ]
        for a in attributes:
            for s in self.samples:
                if s['disk_interface'] != 'sata':
                    continue
                if s['disk_type'] != 'ssd' and 'ssd_only' in a['flags']:
                    continue
                if s['disk_type'] != 'hdd' and 'hdd_only' in a['flags']:
                    continue
                with self.subTest(name=s['name'], attrib=a):
                    (regex, group) = self.get_regex_from_template(a['name'])
                    try:
                        # print(a['name'], s['name'])
                        m = re.search(regex, s['text'])
                        value = int(m.group(group))
                        self.assertGreaterEqual(value, 0)
                        # self.assertLessEqual(value, 100)
                    except AttributeError:
                        print("no attribute {} found on {}".format(
                            a['name'], s['name']))


discovery-scripts/windows/smartctl-disks-discovery.ps1
# VERSION = 1.4
$smartctl = "$Env:Programfiles\smartmontools\bin\smartctl.exe"

if ((Get-Command $smartctl -ErrorAction SilentlyContinue) -eq $null) 
{ 
   write-host "Unable to find smartctl. Check that smartmontools package is installed"
   exit
}


$idx = 0
$global_serials
$smart_scanresults = & $smartctl "--scan-open"
$smart_scanresults += & $smartctl "--scan-open" '-dnvme'

$json = ""

write-host "{"
write-host " `"data`":["
foreach ($smart_scanresult in $smart_scanresults)
{
    
    $idx++;
    $disk_args = ""
    $disk_name = ""
    $disk_type = "2"
    $disk_model = ""
    $disk_sn = ""

    if ($smart_scanresult -match '(-d) ([A-Za-z0-9,\+]+)'){
        $disk_args = $matches[1]+$matches[2]
    }
    
    $disk_name = $smart_scanresult.Substring(0,$smart_scanresult.IndexOf(" "))
    $line = & $smartctl "-i" $disk_name $disk_args
    
    #smart enabled?
    $smart_enabled = $line | select-string "SMART.+Enabled$"
    if($smart_enabled) {
        $smart_enabled = 1
    } else {
        $smart_enabled = 0
    }

    if($disk_args -like "*nvme*" -or $disk_name -like "*nvme*" ) {
        $disk_type = "1"
        $smart_enabled = 1
    }


    # Device sn
    $sn = $line | select-string "serial number:"
    $sn = $sn -ireplace "serial number:"

    # Device Model
    $model = [string] ""
    $model= $line | select-string "Device Model:"
    $model=$model -replace "Device Model:"
    if ($model)
    {
        $disk_model=$model.trim()
    }
    # Model Number
    $model = [string] ""
    $model= $line | select-string "Model Number:"
    $model=$model -replace "Model Number:"
    if ($model)
    {
        $disk_model=$model.trim()
    }
    # Device Model(for SAS)
    $model= $line | select-string "Vendor:"
    $model=$model -replace "Vendor:"
    if ($model)
    {
        $disk_model=$model.trim()    
    }
    $model= $line | select-string "Product:"
    $model=$model -replace "Product:"
    if ($model)
    {
        $disk_model+=" "+$model.trim()
    }
    
    
    # Is it HDD, SSD/NVMe or ODD
    # The SMART Values for HDD/SSD are different sometimes
    # I have 2 Discovery-Rules with Filtering 0 or 1.                       
    # 0 is for HDD
    # 1 is for SSD/NVMe
    # 2 is for ODD and will be ignored
    $Drive = $line  | select-string "Rotation Rate:" 
    if($Drive -like "*Solid State Device*") {$disk_type = "1"}
    elseif ($Drive -like "*rpm*") {$disk_type = "0"}
    
    if($disk_type -eq "2") { 
        #Can't determine, lets go extended
        $extended_line = & $smartctl "-a" $disk_name $disk_args
        
        #search for Spin_Up_Time or Spin_Retry_Count
        if ($extended_line  | select-string "Spin_" -CaseSensitive){
            $disk_type = "0"
        }
        #search for SSD in uppercase
        elseif ($extended_line  | select-string " SSD " -CaseSensitive){
            $disk_type = "1"
        }
        #search for NVMe
        elseif ($extended_line  | select-string "NVMe" -CaseSensitive){
            $disk_type = "1"
            $smart_enabled = "1"
        }
        elseif ($extended_line | select-string "177 Wear_Leveling" -CaseSensitive) {
            $disk_type = "1"
        }
        elseif ($extended_line | select-string "231 SSD_Life_Left" -CaseSensitive) {
            $disk_type = "1"
        }
        elseif ($extended_line | select-string "233 Media_Wearout_" -CaseSensitive) {
            $disk_type = "1"
        }
    }

    if ($sn) {
        $disk_sn=$sn.trim()
    
        if ($global_serials -contains $disk_sn ){
            continue
            #skip duplicated disk, go to next one
        } else {
            #add only smart capable disks to global serials
            if ($smart_enabled -eq 1){
                $global_serials+=$disk_sn
            }
        }
    }

    if ($sn) {
        $disk_sn=$sn.trim()
    
        if ($global_serials -contains $disk_sn ){
            continue
            #skip duplicated disk, go to next one
        } else {
            #add only smart capable disks to global serials
            if ($smart_enabled -eq 1){
                $global_serials+=$disk_sn
            }
        }
    }
              
    if ($idx -eq 1)
    {
        
    } else
    {
        $json +=  ",`n"
    }
    
    $json += "`t {`n " +
            "`t`t`"{#DISKSN}`":`""+$disk_sn+"`""+ ",`n" +        
            "`t`t`"{#DISKMODEL}`":`""+$disk_model+"`""+ ",`n" +        
            "`t`t`"{#DISKNAME}`":`""+$disk_name+"`""+ ",`n" +
            "`t`t`"{#DISKCMD}`":`""+$disk_name+" "+$disk_args+"`"" +",`n" + 
            "`t`t`"{#SMART_ENABLED}`":`""+$smart_enabled+"`"" +",`n" +
            "`t`t`"{#DISKTYPE}`":`""+$disk_type+"`"" +"`n" +
           "`t }"    

    
    
}
write-host $json
write-host " ]"
write-host "}"


discovery-scripts/nix/smartctl-disks-discovery.pl
#!/usr/bin/perl
use warnings;
use strict;

#must be run as root
my $VERSION = 1.4;

#add path if needed into $smartctl_cmd
my $smartctl_cmd = "/usr/sbin/smartctl";
die "Unable to find smartctl. Check that smartmontools package is installed.\n" unless (-x $smartctl_cmd);
my $sg_scan_cmd = "/usr/bin/sg_scan";
my $nvme_cmd = "/usr/sbin/nvme";
my @input_disks;
my @global_serials;
my @smart_disks;

# by providing additional positional arguments you can add disks that are hardly discovered otherwise.
# for example to force discovery of the following disks provide '/dev/sda_-d_sat+megaraid,00 /dev/sda_-d_sat+megaraid,01' as arguments
if (@ARGV>0) {
    foreach my $disk_line (@ARGV) {
        $disk_line =~ s/_/ /g;
        my ($disk_name) = $disk_line =~ /(\/(.+?))(?:$|\s)/;
        my ($disk_args) = $disk_line =~ /(-d [A-Za-z0-9,\+]+)/;
        if (!defined($disk_args)) {
           $disk_args = '';
        }


        if ( $disk_name and defined($disk_args) ) {
            push @input_disks,
                {
                    disk_name => $disk_name,
                    disk_args => $disk_args
                };
        }
    }
}


if ( $^O eq 'darwin' ) {    # if MAC OSX (limited support, consider to use smartctl --scan-open)

    while ( glob('/dev/disk*') ) {
        if ( $_ =~ /\/(disk+[0-9])$/ ) { 
            push @input_disks,
              {
                disk_name => $1,
                disk_args => ''
              };
        }
    }
}
else {
    for (`$smartctl_cmd --scan-open`.
         `$smartctl_cmd --scan-open -dnvme`) {

        #my $testline = "# /dev/sdc -d usbjmicron # /dev/sdc [USB JMicron], ATA device open" ;
        #for ($testline) {
        #splitting lines like  "/dev/sda -d scsi # /dev/sda, SCSI device"
        #"/dev/sda [SAT] -d sat [ATA] (opened)" # in debian 6 and smartctl 5.4
        #"/dev/sda -d sat # /dev/sda [SAT], ATA device" # in debian 8 and smartctl 6.4
        #"/dev/bus/0 -d megaraid,01" for megaraid
        #"# /dev/sdc -d usbjmicron # /dev/sdc [USB JMicron], ATA device open "

        my ($disk_name) = $_ =~ /(\/(.+?))(?:$|\s)/;
        my ($disk_args) = $_ =~ /(-d [A-Za-z0-9,\+]+)/;

        if ( $disk_name and $disk_args ) {
            push @input_disks,
              {
                disk_name => $disk_name,
                disk_args => $disk_args
              };
        }

    }

    if (-x $sg_scan_cmd){
        foreach my $line (`$sg_scan_cmd -i`) {
            ## sg_scan -i
            # https://github.com/v-zhuravlev/zbx-smartctl/pull/29
            #/dev/sg0: scsi0 channel=0 id=0 lun=0
            #    ATA       TOSHIBA MG03ACA1  FL1D [rmb=0 cmdq=1 pqual=0 pdev=0x0] 
            #/dev/sg1: scsi0 channel=1 id=0 lun=0
            #    Dell      Virtual Disk      1028 [rmb=0 cmdq=1 pqual=0 pdev=0x0] 
            #/dev/sg2: scsi0 channel=0 id=1 lun=0
            #    ATA       TOSHIBA MG03ACA1  FL1D [rmb=0 cmdq=1 pqual=0 pdev=0x0]
            if ($line =~ /(\/(.+?)):/){
                    my ($disk_name) = $1;
                    my ($disk_args) = "";

                    push @input_disks,
                        {
                            disk_name => $disk_name,
                            disk_args => $disk_args
                        };
            }

        }
    }

    if (-x $nvme_cmd){
        foreach my $line (`$nvme_cmd list`) {
            ## sg_scan -i
            # Node             SN                   Model                                    Namespace Usage                      Format           FW Rev
            # ---------------- -------------------- ---------------------------------------- --------- -------------------------- ---------------- --------
            # /dev/nvme0n1     S3W8NX0M10ZZZZ       SAMSUNG MZVLB512HAJQ-00000               1          18.87  GB / 512.11  GB    512   B +  0 B   EXA7301Q
            # /dev/nvme1n1     S3W8NX0M15ZZZZ       SAMSUNG MZVLB512HAJQ-00000               1         511.77  GB / 512.11  GB    512   B +  0 B   EXA7301Q
            if ($line =~ /(\/(?:.+?))\s/){
                    my ($disk_name) = $1;
                    my ($disk_args) = "";

                    push @input_disks,
                        {
                            disk_name => $disk_name,
                            disk_args => $disk_args
                        };
            }

        }
    }
}

foreach my $disk (@input_disks) {

    my @output_arr;
    #initialize disk defaults:
    $disk->{disk_model}='';
    $disk->{disk_sn}='';
    $disk->{subdisk}=0;
    $disk->{disk_type}=2; # other

    if ( @output_arr = get_smart_disks($disk) ) {
        push @smart_disks, @output_arr;
    }

}

json_discovery( \@smart_disks );

sub get_smart_disks {
    my $disk = shift;
    my @disks;

    $disk->{smart_enabled} = 0;

    chomp( $disk->{disk_name} );
    chomp( $disk->{disk_args} );
    
    $disk->{disk_cmd} = $disk->{disk_name};
    if (length($disk->{disk_args}) > 0){
        $disk->{disk_cmd}.=q{ }.$disk->{disk_args};
        if ( $disk->{subdisk} == 1 and $disk->{disk_args} =~ /-d\s+[^,\s]+,(\S+)/) {
            $disk->{disk_name} .= " ($1)";
        }
    }

    #my $testline = "open failed: Two devices connected, try '-d usbjmicron,[01]'";
    #my $testline = "open device: /dev/sdc [USB JMicron] failed: Two devices connected, try '-d usbjmicron,[01]'";
    #if ($disk->{subdisk} == 1) {
    #$testline = "/dev/sdb -d usbjmicron,$disk->{disk_args} # /dev/sdb [USB JMicron], ATA device";
    #}
    my @smartctl_output = `$smartctl_cmd -i $disk->{disk_cmd} 2>&1`;
    foreach my $line (@smartctl_output) {
        #foreach my $line ($testline) {
        #print $line;
        if ( $line =~ /^SMART.+?: +(.+)$/ ) {

            if ( $1 =~ /Enabled/ ) {
                $disk->{smart_enabled} = 1;
            }
            #if SMART is disabled then try to enable it (also offline tests etc)
            elsif ( $1 =~ /Disabled/ ) {
                foreach (`smartctl -s on -o on -S on $disk->{disk_cmd}`)
                {
                    if (/SMART Enabled/) { $disk->{smart_enabled} = 1; }
                }
            }
        }
    }
    my $vendor = '';
    my $product = '';
    foreach my $line (@smartctl_output) {
        # SAS: filter out non-disk devices (enclosure, cd/dvd)
        if ( $line =~ /^Device type: +(.+)$/ ) {
                if ( $1 ne "disk" ) {
                    return;
                }
        }
        # Areca: filter out empty slots
        if ( $line =~ /^Read Device Identity failed: empty IDENTIFY data/ ) {
            return;
        }
    
    
        
        if ( $line =~ /^serial number: +(.+)$/i ) {
                $disk->{disk_sn} = $1;
        }
        if ( $line =~ /^Device Model: +(.+)$/i ) {
                $disk->{disk_model} = $1;
        }

        #for NVMe disks and some ATA: Model Number:
        if ( $line =~ /^Model Number: +(.+)$/i ) {
                $disk->{disk_model} = $1;
        }
        #for SAS disks: Model = Vendor + Product
        elsif ( $line =~ /^Vendor: +(.+)$/ ) {
                $vendor = $1;
                $disk->{disk_model} = $vendor;
        }
        elsif ( $line =~ /^Product: +(.+)$/ ) {
                $product = $1;
                $disk->{disk_model} .= q{ }.$product;
        }
        
        if ( $line =~ /Rotation Rate: (.+)/i ) {

            if ( $1 =~ /Solid State Device/i ) {
                $disk->{disk_type} = 1;
            }
            elsif( $1 =~ /rpm/ ) {
                $disk->{disk_type} = 0;
            }
        }

        if ( $line =~ /Permission denied/ ) {

            warn $line;

        }
        elsif ( $disk->{subdisk} == 0 and $line =~ /failed: [A-zA-Z]+ devices connected, try '(-d [a-zA-Z0-9]+,)\[([0-9]+)\]'/) {
            #check for usbjmicron: "open failed: Two devices connected, try '-d usbjmicron,[01]'"
            # or "open device: /dev/sdc [USB JMicron] failed: Two devices connected, try '-d usbjmicron,[01]'"
            #not $disk->{subdisk} works as a guard against endless recursion

            foreach ( split //, $2 ) {    #splitting [01]

                push @disks,
                  get_smart_disks(
                    {
                        disk_name => $disk->{disk_name},
                        disk_args => $1 . $_,
                        subdisk   => 1
                    }
                  );

            }
            return @disks;

        }
        
    }
    
    # Areca SATA RAID
    if ( $disk->{subdisk} == 0 and $vendor eq "Areca" and $product eq "RAID controller" ) {
        for (my $i = 1; $i <= 16; $i++) {
            push @disks,
                get_smart_disks(
                    {
                        disk_name => $disk->{disk_name},
                        disk_args => "-d areca,$i",
                        disk_model => '',
                        disk_sn => '',
                        subdisk   => 1
                    }
                );
        }
        return @disks;
    }

    if ( $disk->{disk_name} =~ /nvme/ or $disk->{disk_args} =~ /nvme/) {
            $disk->{disk_type} = 1; # /dev/nvme is always SSD
            $disk->{smart_enabled} = 1;
    }
    # if disk_type is still unknown after parsing then rerun with extended -a:
    if ( $disk->{disk_type} == 2) {
            foreach my $extended_line (`$smartctl_cmd -a $disk->{disk_cmd} 2>&1`){

                #search for Spin_Up_Time or Spin_Retry_Count
                if ($extended_line  =~ /Spin_/){
                    $disk->{disk_type} = 0;
                    last;
                }
                #search for SSD in uppercase
                elsif ($extended_line  =~ / SSD /){
                    $disk->{disk_type} = 1;
                    last;
                }
                #search for NVME
                elsif ($extended_line  =~ /NVMe/){
                    $disk->{disk_type} = 1;
                    last;
                }
                #search for 177 Wear Leveling Count(present only on SSD):
                elsif ($extended_line  =~ /177 Wear_Leveling/){
                    $disk->{disk_type} = 1;
                    last;
                }
                #search for 231 SSD_Life_Left (present only on SSD)
                elsif ($extended_line  =~ /231 SSD_Life_Left/){
                    $disk->{disk_type} = 1;
                    last;
                }
                #search for 233 Media_Wearout_Indicator (present only on SSD)
                elsif ($extended_line  =~ /233 Media_Wearout_/){
                    $disk->{disk_type} = 1;
                    last;
                }
            }
    }
    
    # push to global serials list
    if ($disk->{smart_enabled} == 1){
        #print "Serial number is ".$disk->{disk_sn}."\n";
        if ( grep /$disk->{disk_sn}/, @global_serials ) {
            # print "disk already exists skipping\n";
            return;
        }
        push @global_serials, $disk->{disk_sn};
    }

    push @disks, $disk;
    return @disks;

}



sub escape_json_string {
    my $string = shift;
    $string =~  s/(\\|")/\\$1/g;
    return $string;
}

sub json_discovery {
    my $disks = shift;

    my $first = 1;
    print "{\n";
    print "\t\"data\":[\n\n";

    foreach my $disk ( @{$disks} ) {

        print ",\n" if not $first;
        $first = 0;
        print "\t\t{\n";
        print "\t\t\t\"{#DISKMODEL}\":\"".escape_json_string($disk->{disk_model})."\",\n";
        print "\t\t\t\"{#DISKSN}\":\"".escape_json_string($disk->{disk_sn})."\",\n";
        print "\t\t\t\"{#DISKNAME}\":\"".escape_json_string($disk->{disk_name})."\",\n";
        print "\t\t\t\"{#DISKCMD}\":\"".escape_json_string($disk->{disk_cmd})."\",\n";
        print "\t\t\t\"{#SMART_ENABLED}\":\"".$disk->{smart_enabled}."\",\n";
        print "\t\t\t\"{#DISKTYPE}\":\"".$disk->{disk_type}."\"\n";
        print "\t\t}";

    }
    print "\n\t]\n";
    print "}\n";

}
